Add extra field in kill container json

devimc opened the pull request:
This patch adds an extra field in kill container json
in order to make hyperstart compatible with docker 1.13.x.
"allProcesses" is a boolean field used to specify if
the signal must be send to all running processes inside
the container

Signed-off-by: Julio Montes <julio.montes@intel.com>

sboeuf commented:
@devimc code looks good, but I wonder if we need this. Indeed, you can send several SignalProcess() to do the same.

amshinde commented on src/init.c:
```diff
@@ -618,7 +618,28 @@ static int hyper_kill_container(char *json, int length)
+	/* json_object_get_boolean returns '1' only when
+	 * 'allProcesses' exists and is true
+	 */
+	if (1 == all_processes) {
```
Prefer all_processes ==1 check, also for consistency.

devimc replied:
for security reasons I'd prefer to use ```if (1 == all_processes)```
because if I make a mistake like this ```if (1 = all_processes)``` compiler will fail
otherwise with ```if (all_processes = 1)``` condition will be always true

amshinde replied:
@devimc Most modern compilers will throw an error while using an assignment as a truth value.

devimc replied:
@amshinde yes, but only if you have enabled the right flags

grahamwhaley replied:
@devimc the 'const == var' idiom is a valid one (but can cause almost as much debate 'as which editor do you use?'), but as @amshinde notes, it does not seem to be the defacto style in this file, so in this case I think we should stick with 'var == 1'...

devimc commented:
@sboeuf  is there a function to list all container processes?

amshinde commented:
minor nitpick, otherwise
lgtm

devimc commented:
@sboeuf unfortunately hyperstart 0.7.0 has not this functionality (SignalProcess)

dlespiau commented:
@sboeuf the `--all` flag kills all processes inside the container, not just the initial workloads + potential execs (those initial process may have spawned other processes), so it's probably a good idea if it's done from inside the VM. (see https://github.com/opencontainers/runc/commit/e58671e530eb04a233d953c6a56e56804703f80c)

There are quite a few behavioural details, some we maybe be able to get right, some not really:
- Kill all pids, not just initial workload + exec (docker gets the list of pids from the cgroup API)
- From a quick glance, seems runc sends SIGKILL to all processes if the container shares the same pidns as the host, but not if the initial workload is in its own ns, right now, we always have our own pidns, but the initial wokload isn't pid 1 (yes, that's a bug)
- docker seems to wait for the process when killing them: https://github.com/opencontainers/runc/blob/master/libcontainer/init_linux.go#L435
- ...

Currently it seems this commit doesn't go through the list of pids in the ns, which it should probably do, except maybe killing pid 1 hyperstart before we fix that bug.

That said, this commit is also probably enough to unlock a good enough version of `cc-oci-runtime kill --all`, where those edge cases don't matter too much I believe.

dlespiau commented:
Ah, the reason why runc doesn't send SIGKILL to all processes when the container has its own ns is:

>  If the "init" process of a PID namespace terminates, the kernel
>        terminates all of the processes in the namespace via a SIGKILL
>        signal.  This behavior reflects the fact that the "init" process is
>        essential for the correct operation of a PID namespas

From: http://man7.org/linux/man-pages/man7/pid_namespaces.7.html

devimc commented:
Hi @amshinde @sboeuf @dlespiau 
could you please review it again?

jodh-intel commented on src/init.c:
```diff
@@ -618,7 +618,41 @@ static int hyper_kill_container(char *json, int length)
+	const int all_processes = json_object_get_boolean(json_object(value),
+		"allProcesses");
+
+	struct hyper_exec *exec;
```
Couldn't this be defined inside the `if` block below?

devimc replied:
Hi @jodh-intel 
yep, fixing it
Thank you!

jodh-intel commented:
lgtm

dlespiau commented:
If I read things correctly you are relying on the fact the processes inside the pid ns will get killled if the local pid 1 goes away. That should really be part of the commit message/code or both.

With that:
lgtm.

Now, this is not the final solution we want, we also want to get rid of the hyperstart process as pid 1, that should be a separate problem but now it's not. Fortunately that shouldn't be too bad to fix when someone looks at the problem.

dlespiau commented:
@devimc You have marked that PR as `[WIP]`, have you found a problem with it or intend to rework it? FWIW, I'm fine with merging that version and we can come up later on with a better fix on top of fixing hyperstart being pid 1 in the container.

devimc commented:
@dlespiau , yep I found a problem in the code, seems like hyperstart kills all remaining processes when the 'workload' ends [1] hence we only have to send the signal to exec processes (kill --all functionality) and ```hyper_kill_container_processes``` will do the magic

[1] https://github.com/clearcontainers/hyperstart/blob/0.7.0-clearcontainers/src/exec.c#L746

dlespiau commented:
ah, of course they use the racy approach we've talked about, without cgroup freezing it's hard to do otherwise. Good enough I guess :)

Commit:
Add extra field in kill container json

This patch adds an extra field in kill container json
in order to make hyperstart compatible with docker 1.13.x.
"allProcesses" is a boolean field used to specify if
the signal must be send to all running processes inside
the container

Signed-off-by: Julio Montes <julio.montes@intel.com>

devimc commented:
@dlespiau @jodh-intel 
PR fixed
Thank you!

dlespiau commented:
Ok, still looking good. Let's merge this.
